src/flashcart/ed64/ed64_ll.c

```diff
@@ -0,0 +1,409 @@
+    // We do this because 128Kib is not recognised and a 32KiB allocates too little 
+    pi_dma_from_sram(buffer, -(size - KiB(16)), size);
+
+    dma_wait();
+
+    io_write(PI_BSD_DOM2_LAT_REG, 0x40);
+    io_write(PI_BSD_DOM2_PWD_REG, 0x12);
+    io_write(PI_BSD_DOM2_PGS_REG, 0x07);
+    io_write(PI_BSD_DOM2_RLS_REG, 0x03);
+
+    return 1;
+
+}
+
+
+int ed64_ll_get_sram (uint8_t *buffer, int size) {
+
+    dma_wait();
+
+    io_write(PI_BSD_DOM2_LAT_REG, 0x05);
+    io_write(PI_BSD_DOM2_PWD_REG, 0x0C);
+    io_write(PI_BSD_DOM2_PGS_REG, 0x0D);
+    io_write(PI_BSD_DOM2_RLS_REG, 0x02);
+
+    dma_wait();
+
+    pi_initialize();
+
+    dma_wait();
+
+    pi_dma_from_sram(buffer, 0, size) ;
+
+    dma_wait();
+
+    io_write(PI_BSD_DOM2_LAT_REG, 0x40);
+    io_write(PI_BSD_DOM2_PWD_REG, 0x12);
+    io_write(PI_BSD_DOM2_PGS_REG, 0x07);
+    io_write(PI_BSD_DOM2_RLS_REG, 0x03);
+
+    return 1;
+
+}
+
+int ed64_ll_get_eeprom (uint8_t *buffer, int size) {
+
+    int blocks=size/8;
+    for( int b = 0; b < blocks; b++ ) {
+        eeprom_read( b, &buffer[b * 8] );
+    }
+
+    return 1;
+
+}
+
+
+int ed64_ll_get_fram (uint8_t *buffer, int size) {
+
+    ed64_ll_set_save_type(SAVE_TYPE_SRAM_128K); //2
+    dma_wait();
+
+    ed64_ll_get_sram_128(buffer, size);
+    data_cache_hit_writeback_invalidate(buffer, size);
+
+    dma_wait();
+    ed64_ll_set_save_type(SAVE_TYPE_FLASHRAM);
+
+    return 1;
+
+}
+
+/*
+sram upload
+*/
+
+
+int ed64_ll_set_sram_128 (uint8_t *buffer, int size) {
+
+    //half working
+    dma_wait();
+    //Timing
+    pi_initialize_sram();
+
+    //Readmode
+    pi_initialize();
+
+    data_cache_hit_writeback_invalidate(buffer,size);
+    dma_wait();
+
+    // Offset Large RAM size 128KiB with a 16KiB nudge to allocate enough space
+    // We do this because 128Kib is not recognised and a 32KiB allocates too little 
+    pi_dma_to_sram(buffer, -(size - KiB(16)), size);
+    data_cache_hit_writeback_invalidate(buffer,size);
+
+    //Wait
+    dma_wait();
+    //Restore evd Timing
+    ed64_ll_set_sdcard_timing();
+
+    return 1;
+
+}
+
+
+int ed64_ll_set_sram (uint8_t *buffer, int size) {
+
+    //half working
+    dma_wait();
+    //Timing
+    pi_initialize_sram();
+
+    //Readmode
+    pi_initialize();
+
+    data_cache_hit_writeback_invalidate(buffer,size);
+    dma_wait();
+
+    pi_dma_to_sram(buffer, 0, size);
+    data_cache_hit_writeback_invalidate(buffer,size);
+
+    //Wait
+    dma_wait();
+    //Restore evd Timing
+    ed64_ll_set_sdcard_timing();
+
+    return 1;
+
+}
+
+
+int ed64_ll_set_eeprom (uint8_t *buffer, int size) {
+
+    int blocks=size/8;
+    for( int b = 0; b < blocks; b++ ) {
+        eeprom_write( b, &buffer[b * 8] );
+    }
+
+    return 1;
+
+}
+
+int ed64_ll_set_fram (uint8_t *buffer, int size) {
+
+    ed64_ll_set_save_type(SAVE_TYPE_SRAM_128K);
+    dma_wait();
+
+    ed64_ll_set_sram_128(buffer, size);
+    data_cache_hit_writeback_invalidate(buffer, size);
```

Comment:
Cache writeback is in wrong place, it should happen before doing write in the `ed64_ll_set_sram_128` function. This applies to all functions above.